homeassistant/components/sensor/nilu_air_quality.py

```diff
@@ -0,0 +1,238 @@
+ATTR_POLLUTION_INDEX = "pollution_index"
+ATTR_MAX_POLLUTION_INDEX = 'max_pollution_index'
+
+CONF_ALLOWED_AREAS = [
+    'Ålesund',
+    'Zeppelinfjellet',
+    'Tustervatn',
+    'Trondheim',
+    'Tromsø',
+    'Sør-Varanger',
+    'Stavanger',
+    'Sarpsborg',
+    'Sandve',
+    'Prestebakke',
+    'Oslo',
+    'Narvik',
+    'Moss',
+    'Mo i Rana',
+    'Lørenskog',
+    'Lillestrøm',
+    'Lillesand',
+    'Lillehammer',
+    'Kårvatn',
+    'Kristiansand',
+    'Karasjok',
+    'Hurdal',
+    'Harstad',
+    'Hamar',
+    'Halden',
+    'Grenland',
+    'Gjøvik',
+    'Fredrikstad',
+    'Elverum',
+    'Drammen',
+    'Bærum',
+    'Brumunddal',
+    'Bodø',
+    'Birkenes',
+    'Bergen'
+]
+
+CONF_ALLOWED_MEASURABLE_COMPONENTS = [
+    'index',
+    'PM1',
+    'PM10',
+    'PM2.5',
+    'NO',
+    'NO2',
+    'NOx',
+    'O3',
+    'CO',
+    'SO2'
+]
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Inclusive(CONF_LATITUDE, 'coordinates',
+                  'Latitude and longitude must exist together'): cv.latitude,
+    vol.Inclusive(CONF_LONGITUDE, 'coordinates',
+                  'Latitude and longitude must exist together'): cv.longitude,
+    vol.Required(CONF_MONITORED_CONDITIONS):
+        vol.All(cv.ensure_list, [vol.In(CONF_ALLOWED_MEASURABLE_COMPONENTS)]),
+    vol.Exclusive(CONF_AREA, 'station_collection',
+                  'Can only configure one specific station or '
+                  'stations in a specific area pr sensor. '
+                  'Please only configure station or area.'):
+        vol.All(cv.string, vol.In(CONF_ALLOWED_AREAS)),
+    vol.Exclusive(CONF_STATION, 'station_collection',
+                  'Can only configure one specific station or '
+                  'stations in a specific area pr sensor. '
+                  'Please only configure station or area.'):
+        cv.string,
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    vol.Optional(CONF_SHOW_ON_MAP, default=False): cv.boolean,
+})
+
+
+async def async_setup_platform(hass, config, add_entities,
+                               discovery_info=None):
+    """Set up the NILU air quality sensors."""
+    import niluclient as nilu
+    name = config.get(CONF_NAME)
+    area = config.get(CONF_AREA)
+    station = config.get(CONF_STATION)
+    monitored_conditions = config.get(CONF_MONITORED_CONDITIONS)
+    show_on_map = config.get(CONF_SHOW_ON_MAP)
+
+    sensors = []
+
+    if station:
+        stations = [station]
+    elif area:
+        stations = nilu.lookup_stations_in_area(area)
+    else:
+        latitude = config.get(CONF_LATITUDE, hass.config.latitude)
+        longitude = config.get(CONF_LONGITUDE, hass.config.longitude)
+        location_client = nilu.create_location_client(latitude, longitude)
+        stations = location_client.station_names
+
+    for station in stations:
+        client = NiluData(nilu.create_station_client(station))
+        client.update()
```

Comment:
Does this do I/O? We're in async context here so are not allowed to do regular I/O.

Reply:
Yes, that one does I/O, so I have changed to sync. I'm used to having async as desired when doing I/O, so didn't come to mind that it might be wrong usage.